api/websocket/server/relay.go

```diff
@@ -124,11 +125,44 @@ func (ws *WsServer) sendInboundRelayMessage(relayMessage *pb.Relay) {
 				sigChainLen: int(relayMessage.SigChainLen),
 			})
 		}
-	} else {
+		if time.Duration(relayMessage.MaxHoldingSeconds) > pongTimeout/time.Second {
+			ok := ws.messageDeliveredCache.Push(relayMessage)
+			if !ok {
+				log.Warningf("MessageDeliveredCache full, discarding messages.")
+			}
+		}
+	} else if relayMessage.MaxHoldingSeconds > 0 {
 		ws.messageBuffer.AddMessage(clientID, relayMessage)
 	}
 }
 
+func (ws *WsServer) startCheckingLostMessages() {
+	for {
```

Comment:
Should we need something interval for each loop in order to avoid eat all cpu.idle?

Reply:
It won't. Pop will block until the internal channel is available: https://github.com/nknorg/nkn/pull/685/files#diff-6bfb9c74eef98d9279404e3f0e69f0ffR39

Reply:
I know Pop will block if channel is empty. I means that the popped msg still can't deliver and will be pushed back again very quickly. The channel is not "empty" but pop/push same msg repeatedly with high frequency.

Reply:
Popped msg that can't deliver will be put back into message buffer, but messageDeliveredCache, so it won't happen